File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

classify :: Int -> Maybe Classification
classify n
  | n < 1 = Nothing
  | factorsum == n = Just Perfect
  | factorsum > n = Just Abundant
  | factorsum < n = Just Deficient
  where
    factorsum = sum $ factors n

factors :: Int -> [Int]
factors n = filter factorOfN [1..(n - 1)]
  where
    factorOfN x = n `mod` x == 0

File: solutions/scala/perfect-numbers/1/src/main/scala/PerfectNumbers.scala
object PerfectNumbers {
  def classify(n: Int): Either[String, NumberType] = {
    if(n > 0) {
      val sum = factors(n).sum
      if(sum == n) {
        Right(NumberType.Perfect)
      } else if (sum > n) {
        Right(NumberType.Abundant)
      } else {
        Right(NumberType.Deficient)
      }
    } else {
      Left("Classification is only possible for natural numbers.")
    }
  }

  def factors(n: Int): Seq[Int] = {
    (1 until n).filter(n % _ == 0)
  }
}

enum NumberType {
  case Abundant, Perfect, Deficient
}